cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+    """
+    Create, update, or replace a component in a library based on a source key and OLX.
+
+    (We assume that the destination is a library rather than some other future kind of learning
+     package, but let's keep than an internal assumption.)
+    """
+    component_type = authoring_api.get_or_create_component_type("xblock.v1", source_key.block_type)
+
+    target_key = _deduplicate_block_id(
+        context,
+        source_key,
+        component_type,
+        olx,
+    )
+
+    try:
+        log.info(f"Attempting to retrieve component for {target_key}")
+        component = authoring_api.get_components(context.target_package_id).get(
+            component_type=component_type,
+            local_key=target_key.block_id,
+        )
+        component_existed = True
+        # Do we have a specific method for this?
+        component_deleted = not component.versioning.draft
+        log.info(f"It's existed: {component_existed}")
+        log.info(f"It's deleted: {component_deleted}")
+    except Component.DoesNotExist:
+        component_existed = False
+        component_deleted = False
+        try:
+            libraries_api.validate_can_add_block_to_library(
+                context.target_library_key, target_key.block_type, target_key.block_id
+            )
+        except libraries_api.IncompatibleTypesError as e:
+            log.error(f"Error validating block for library {context.target_library_key}: {e}")
+            return None
+        component = authoring_api.create_component(
+            context.target_package_id,
+            component_type=component_type,
+            local_key=target_key.block_id,
+            created=context.created_at,
+            created_by=context.created_by,
+        )
+
+    # Component existed and we do not replace it and it is not deleted previously
+    if component_existed and not component_deleted and not context.replace_existing:
+        return component.versioning.draft.publishable_entity_version
+
+    # If component existed and was deleted or we have to replace the current version
+    # Create the new component version for it
+    component_version = libraries_api.set_library_block_olx(target_key, new_olx_str=olx)
+    for filename, content_pk in content_by_filename.items():
+        filename_no_ext, _ = os.path.splitext(filename)
+        if filename_no_ext not in olx:
+            continue
+        new_path = f"static/{filename}"
+        _create_component_version_content_safely(
+            component_version_pk=component_version.pk,
+            content_pk=content_pk,
+            key=new_path
+        )
+    return component_version.publishable_entity_version
+
+
+def _deduplicate_container(
+    context: MigrationContext,
+    source_key: UsageKey,
+    container_type: ContainerType,
+    title: str,
+) -> LibraryContainerLocator:
+    if not source_key or not source_key.block_id:
+        raise ValueError(f"Invalid source_key: {source_key}")
+
```

Comment:
```suggestion
```
Since we're using type annotations, mypy should forbid this from ever happening, so I think it we can remove this case. Unless you were seeing it happen in your testing?